src/main/host/syscall/handler/socket.rs

```diff
@@ -665,7 +665,164 @@ fn read_sockaddr(
         )?;
     }
 
-    Ok(Some(nix::sys::socket::sockaddr_storage_to_addr(
-        &addr, addr_len,
-    )?))
+    Ok(Some(sockaddr_storage_to_nix(&addr, addr_len)?))
+}
+
+fn sockaddr_storage_to_nix(
+    addr: &nix::sys::socket::sockaddr_storage,
+    mut addr_len: usize,
+) -> Result<nix::sys::socket::SockAddr, nix::Error> {
+    // Workaround for a nix bug where a function called from within 'sockaddr_storage_to_addr()'
+    // assumes that the given length is the exact length of the address, whereas the documentation
+    // for 'sockaddr_storage_to_addr()' only specifies "[The len] must be at least as large as all
+    // the useful parts of the structure".  For example if a unix address containing 'sockaddr_un {
+    // family: AF_UNIX, path: b"test\0" }' is given to 'sockaddr_storage_to_addr()' with a length of
+    // 'sizeof(sockaddr_un)', it may panic due to an out-of-bounds read. This also applies to
+    // abstract addresses.
+    if addr.ss_family == libc::AF_UNIX as u16 {
+        // copy the bytes from the 'sockaddr_storage' into a 'sockaddr_un'
+        let mut tmp_addr: libc::sockaddr_un = unsafe { std::mem::zeroed() };
+        unsafe {
+            std::ptr::copy(
+                &addr as *const _ as *const u8,
+                &mut tmp_addr as *mut _ as *mut u8,
+                std::mem::size_of_val(&tmp_addr),
+            )
+        };
+
+        // find the position of the first null byte
+        let null_position = tmp_addr.sun_path.iter().position(|c| *c == 0);
+
+        // if there was at least one null byte
+        if let Some(null_position) = null_position {
+            // if a pathname address and not an abstract address
+            if null_position != 0 {
+                // reduce the addr len if possible
+                addr_len = std::cmp::min(
+                    addr_len,
+                    std::mem::size_of_val(&addr.ss_family) + null_position,
+                );
+            }
+        }
+    };
+
+    nix::sys::socket::sockaddr_storage_to_addr(addr, addr_len)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_unix_sockaddr_to_nix_1() {
```

Comment:
Could these names be a little more descriptive about what property their testing? e.g. replace the _1, _2, and _3 suffixes with something about what distinguishes these from eachother?